### PR TITLE
Allow the specification of blocks allowed within columns

### LINKED ISSUE #18161: Individual Column(s) should be able to define allowed blocks
**Is your feature request related to a problem? Please describe.**
Momently, an individual column can not define specific allowed blocks to be added. Furthermore, newly added columns from the slider in the columns settings also can not only allow certain blocks.

Use case - As a WordPress developer I want to be able to assign allowed blocks for individual columns in my custom block I am developing so I can guide and restrict the user in adding certain blocks.

**Describe the solution you'd like**
I want to introduce a new attribute called allowedBlocks, which is used to define the allowed blocks in the column(s). Locally, I have made a solution to this problem which locates to PR #18164.

**Describe alternatives you've considered**
I have considered creating my own column(s) block, which I did not find to be a relevant solution as I think other people could run into this problem as well.

### PR DESCRIPTION
**Pull request by JesserH**
## Description

This an updated version of #25183. 
I and @aristath developed a solution to the issue where an individual column could not define any specifically allowed blocks to be added in it. This also applies to newly added columns in the columns block (from the slider at the columns settings) also specify the blocks allowed to be added into that column. We believe that this is important to custom Gutenberg block development to nest blocks in a good and effective way.

## How has this been tested?

Me and @aristath have tested this locally.

## Screenshots 
Screenshot 2020-09-09 at 12 14 13

## Types of changes

We have added a new attribute called allowedBlocks in the column, and a childAllowedBlocks attribute within the columns block, which allows for defining the allowed blocks per individual column.

Example of use:
```
const allowedBlocks = [
  "core/paragraph",
  "core/separator",
  "core/cover",
];

//the code below can be used in a template
[ [ 'core/columns', { allowedChildBlocks: allowedBlocks }, [
      [ 'core/column', { allowedBlocks }, [
          //here a block is added, but with the allowedBlocks property more specifically allowed blocks can be added
          [ 'cgb/recipe-name-block', { name: attributes.name, allowedBlocks },  ],
      ] ],
      [ 'core/column', { allowedBlocks }, [
        [ 'cgb/recipe-description-block', { recipeDescription: attributes.description, allowedBlocks },  ],          
     ] ],
  ] ] ],

```

## Checklist:
- [X] My code is tested.
- [X] My code follows the WordPress code style. 
- [X] My code follows the accessibility standards. 
- [X] My code has proper inline documentation. 
- [X] I've included developer documentation if appropriate. 
- [X] I've updated all React Native files affected by any refactorings/renamings in this PR. 

fixes #18161

### CODE REVIEW

**Comment by ajitbohra** [comment]
@JesserH this seems to be a duplicate of #25183 which is already merged.

**Comment by aristath** [comment]
Reopening this one.
#25183 appears merged in the Github UI but was not actually merged. There was a git mix-up and a force-push on the master branch just seconds after that PR got merged by mistake. As a result, the PR is marked as merged (even though it wasn't merged) and there is no way to reopen it.

**Comment by aristath** [comment]
Tested and confirmed this works, details on https://github.com/WordPress/gutenberg/pull/25183#issuecomment-689488277

**Comment by raphaelparent** [comment]
Hi! I was wondering if there was any progress on this? Or if it was merged in part of another PR?

The last commits is dated of October 23rd, it was confirmed as working but the merge is blocked because of failed checks. I've looked around quit a bit and this is the most recent and up to date PR I've seen concerning the issue of allowed blocks in column elements.

**Comment by aristath** [comment]
@JesserH could you please sync your fork with the main repo? I believe most of the mailing tests have been fixed since then

File: packages/block-library/src/column/block.json
Comment (aristath) [inline packages/block-library/src/column/block.json:17]: Indentation is off here
```suggestion
		},
```

**Review by aristath — CHANGES_REQUESTED** [review]

**Comment by aristath** [comment]
cc @talldan @ajitbohra

**Comment by talldan** [comment]
I was wondering if it'd be possible to use block context to just declare the attribute once on the parent columns block:
https://developer.wordpress.org/block-editor/developers/block-api/block-context/

The individual column could consume that context and pass it to the inner blocks of the column.

This way there would be a single source of truth for the `allowedBlocks` rather than duplicated attributes and there would then be less risk of them becoming out of sync.

**Comment by talldan** [comment]
Though I've maybe changed my mind looking at how the `templateLock` attribute was defined here - https://github.com/WordPress/gutenberg/pull/26128

It seems like that PR specifically went for a solution where columns could have different template locks. It didn't really tackle whether users could modify the column count after the fact.

Given the main use-case for both would be templating, I wonder if it's worth matching that implementation for consistency as a first step, and then considering how it might work in the columns block as a separate change. What do you think?

**Comment by JesserH** [comment]
@talldan I chose to implement the allowedChildBlocks for the blocks individually because of the columns block which has a slider. To my understanding it's not ideal to pass the value to the new block when the slider is moved upwards (so new blocks spawned).
Additionally, since it is used mostly for templating I think it is valuable to give the creator the customisability over his pre-defined columns, which is only improved functionality if you ask me. 😄 I'd personally prefer to keep the code as it is, but understand if you do want a refactor.
What is your view on this? 😃

**Comment by raphaelparent** [comment]
@aristath thanks for the follow up on this feature!

It seems like the merge is only awaiting your validation for the requested changes. Once it's merge and part of an update (not sure of the flow here) I'll be able to test it and give feedback if I find some problems with it.

**Comment by aristath** [comment]
> @aristath thanks for the follow up on this feature! It seems like the merge is only awaiting your validation for the requested changes. Once it's merged and part of an update (not sure of the flow here) I'll be able to test it and give feedback if I find some problems with it.

I was part of this feature's development so I can't (ethically) review it, I wouldn't be objective. We'll wait for someone else to properly review it when it's time 👍

File: packages/block-library/src/columns/edit.js
Comment (talldan) [inline packages/block-library/src/columns/edit.js:75]: I think this line can be removed, `allowedChildBlocks` isn't a recognized property for inner blocks:
https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/inner-blocks/README.md#props

**Comment by talldan** [comment]
@JesserH From testing, I think `allowedChildBlocks` is misleading, because it sounds like it'd allow an implementor to define the allowed blocks once at the parent `columns` level. I experimented with the following template:
```php
function myplugin_register_book_post_type() {
	$args = array(
		'public'        => true,
		'label'         => 'Books',
		'show_in_rest'  => true,
		'template'      => array(
			array(
				'core/columns',
				array(
					'allowedChildBlocks' => array( 'core/paragraph' ),
				),
				array(
					array(
						'core/column',
					),
					array(
						'core/column',
					),
				),
			),
		),
	);
	register_post_type( 'book', $args );
}
add_action( 'init', 'myplugin_register_book_post_type' );
```

This didn't work, I still need to specify `allowedBlocks` on each column. It looks like all `allowedChildBlocks` does is control the allowed blocks for newly added columns.

It could be renamed, but I think this needs some more thought. I'd be inclined to strip this PR back to just include the changes to the column block `allowedBlock` attribute, and explore the columns block changes in a separate PR.

Also cc'ing @jorgefilipecosta and @mcsf for a review, as it looks like Jorge worked on the template lock for column and Miguel code reviewed.

File: packages/block-library/src/columns/edit.js
Comment (jorgefilipecosta) [reply packages/block-library/src/columns/edit.js:75]: Yes, I agree that this line should be removed.

**Review by jorgefilipecosta — COMMENTED** [review]
I think the allowedBlocks attribute in columns is good addition, it follows the same approach as in https://github.com/WordPress/gutenberg/pull/26128 👍 

Regarding the mechanism to control allowed blocks for newly created columns I agree with @talldan, it seems safer to not include it for now.
I guess if we needed it after we can create something with specific names saying that is for new columns that covers, allowedBlocks and templateLock.

**Comment by JesserH** [comment]
@jorgefilipecosta I updated the PR accordingly 😄

**Comment by talldan** [comment]
@JesserH The changes seem the opposite to what was recommended, I think it'd be best to keep only the changes on the column block.

**Comment by JesserH** [comment]
@talldan Sorry, that's my bad 😅 . I have updated the PR 😄

**Comment by LucaPipolo** [comment]
Any update to this? Would be really great to have this feature! 🙏🏼 😬

**Review by draganescu — APPROVED** [review]
I think this is updated according to the requested changes now, right @talldan?

File: packages/block-library/src/column/edit.js
Comment (gziolo) [inline packages/block-library/src/column/edit.js:75]: What happens when `allowedBlocks` is undefined?

File: packages/block-library/src/column/block.json
Comment (talldan) [inline packages/block-library/src/column/block.json:17]: Would be good to get the white space correct here.

**Comment by aristath** [comment]
Unfortunately we can't push any fixes or update the PR since it's on a fork, and @JesserH has since moved on to other projects.
In order to move this forward, I'm closing this PR and re-submitting it in https://github.com/WordPress/gutenberg/pull/35342
The new PR uses this branch as a starting point, and I simply rebased it, resolved conflicts etc. I'll be addressing the feedback there. Feel free to comment on the new PR, review and push additional changes as needed. The new PR is from the main repo and not a fork so it will be easier to move forward. 👍

File: packages/block-library/src/column/block.json
Comment (aristath) [reply packages/block-library/src/column/block.json:17]: Fixed in the new PR: https://github.com/WordPress/gutenberg/pull/35342 👍

**Pull request closed without merging**
